cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetUpgrader.java

```diff
@@ -93,39 +94,69 @@ private void upgradeSystemDatasets() throws Exception {
-
-    for (HTableDescriptor desc : hAdmin.listTables(userTablePrefix)) {
-      HTableNameConverter hTableNameConverter = new HTableNameConverterFactory().get();
-      TableId tableId = hTableNameConverter.from(desc);
-      LOG.info("Upgrading hbase table: {}, desc: {}", tableId, desc);
-
-      final boolean supportsIncrement = HBaseTableAdmin.supportsReadlessIncrements(desc);
-      final boolean transactional = HBaseTableAdmin.isTransactional(desc);
-      DatasetAdmin admin = new AbstractHBaseDataSetAdmin(tableId, hConf, hBaseTableUtil) {
-        @Override
-        protected CoprocessorJar createCoprocessorJar() throws IOException {
-          return HBaseTableAdmin.createCoprocessorJarInternal(cConf,
-                                                              locationFactory,
-                                                              hBaseTableUtil,
-                                                              transactional,
-                                                              supportsIncrement);
-        }
-
-        @Override
-        protected boolean upgradeTable(HTableDescriptor tableDescriptor) {
-          // we don't do any other changes apart from coprocessors upgrade
-          return false;
-        }
-
-        @Override
-        public void create() throws IOException {
-          // no-op
-          throw new UnsupportedOperationException("This DatasetAdmin is only used for upgrade() operation");
-        }
-      };
-      admin.upgrade();
-      LOG.info("Upgraded hbase table: {}", tableId);
+    for (HTableDescriptor desc : hAdmin.listTables()) {
+      String tableName = desc.getNameAsString();
+      if (!isTableInSystemNamespace(tableName) &&
+           isUserTable(tableName) &&
+           isTableCreatedByCDAP(desc)) {
+        upgradeUserTable(desc);
+      }
     }
   }
 
+
+  private void upgradeUserTable(HTableDescriptor desc) throws IOException {
+    HTableNameConverter hTableNameConverter = new HTableNameConverterFactory().get();
+    TableId tableId = hTableNameConverter.from(desc);
+    LOG.info("Upgrading hbase table: {}, desc: {}", tableId, desc);
+
+    final boolean supportsIncrement = HBaseTableAdmin.supportsReadlessIncrements(desc);
+    final boolean transactional = HBaseTableAdmin.isTransactional(desc);
+    DatasetAdmin admin = new AbstractHBaseDataSetAdmin(tableId, hConf, hBaseTableUtil) {
+      @Override
+      protected CoprocessorJar createCoprocessorJar() throws IOException {
+        return HBaseTableAdmin.createCoprocessorJarInternal(cConf,
+                                                            locationFactory,
+                                                            hBaseTableUtil,
+                                                            transactional,
+                                                            supportsIncrement);
+      }
+
+      @Override
+      protected boolean upgradeTable(HTableDescriptor tableDescriptor) {
+        return false;
```

Comment:
The contract says return true, if HTableDescriptor is modified. Since we don't returning false here.
```
/**
   * Modifies the table descriptor for upgrade.
   *
   * @return true if the table descriptor is modified.
   */
  protected abstract boolean upgradeTable(HTableDescriptor tableDescriptor);
```